bee-tangle/src/vertex.rs

```diff
@@ -1,99 +1,47 @@
-// Copyright 2020 IOTA Stiftung
-//
-// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
-// the License. You may obtain a copy of the License at
-//
-//     http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
-// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and limitations under the License.
+use crate::TransactionRef as TxRef;
 
-use crate::tangle::Tangle;
-
-use bee_transaction::{BundledTransaction as Transaction, Hash};
-
-use std::ops::Deref;
+use bee_transaction::{BundledTransaction as Tx, Hash as TxHash, TransactionVertex};
 
 use async_std::sync::Arc;
-use bitflags::bitflags;
 
-/// A wrapper around `bee_transaction::Transaction` that allows sharing it safely across threads.
 #[derive(Clone)]
-pub struct TransactionRef(Arc<Transaction>);
-
-impl Deref for TransactionRef {
-    type Target = Transaction;
-
-    fn deref(&self) -> &Self::Target {
-        &*self.0
-    }
-}
-
-bitflags! {
-    pub(crate) struct Flags: u8 {
-        const SOLID = 0b0000_0001;
-        const TAIL = 0b0000_0010;
-        const REQUESTED = 0b0000_0100;
-        const MILESTONE = 0b0000_1000;
-    }
-}
-
-pub struct Vertex {
-    id: Hash,
-    inner: TransactionRef,
-    flags: Flags,
+pub struct Vertex<T>
+where
+    T: Clone + Copy,
+{
+    transaction: TxRef,
+    metadata: T,
 }
 
-impl Vertex {
-    pub fn from(transaction: Transaction, hash: Hash) -> Self {
-        let flags = if transaction.is_tail() {
-            Flags::TAIL
-        } else {
-            Flags::empty()
-        };
-
+impl<T> Vertex<T>
+where
+    T: Clone + Copy,
+{
+    pub fn new(transaction: Tx, metadata: T) -> Self {
         Self {
-            id: hash,
-            inner: TransactionRef(Arc::new(transaction)),
-            flags,
+            transaction: TxRef(Arc::new(transaction)),
+            metadata,
         }
     }
 
-    pub fn get_id(&self) -> Hash {
-        self.id
-    }
-
-    pub fn get_ref_to_inner(&self) -> TransactionRef {
-        self.inner.clone()
-    }
-
-    pub fn is_solid(&self) -> bool {
-        self.flags.contains(Flags::SOLID)
-    }
-
-    pub fn set_solid(&mut self) {
-        self.flags.insert(Flags::SOLID);
-    }
-
-    pub fn is_tail(&self) -> bool {
-        self.flags.contains(Flags::TAIL)
+    pub fn get_trunk(&self) -> &TxHash {
+        self.transaction.trunk()
     }
 
-    pub fn is_requested(&self) -> bool {
-        self.flags.contains(Flags::REQUESTED)
+    pub fn get_branch(&self) -> &TxHash {
+        self.transaction.branch()
     }
 
-    pub fn set_requested(&mut self) {
-        self.flags.insert(Flags::REQUESTED);
+    pub fn get_data(&self) -> &TxRef {
```

Comment:
Not sure about the name... It will always be a transaction.

Reply:
I'll change.